### PR TITLE
fix(internals): vercel pkg resolveBinary on Windows

### PR DESCRIPTION
**Pull request by millsp**
Fixes for `prisma@5` with `pkg@5`.

Update 1: In `pkg@5` the paths on Windows don't start with `/snapshot/` anymore, but with `C:\snapshot\` so our logic does not seem to match as it currently checks for `/snapshot/`. So this PR is an attempt to find out if this fixed the said issue - draft for now.

Update 2: This did not fix it on its own. The schema engine is still not found, so there is something else to it. Currently trying to see if that could be a missing annotation for the schema engine. If that does not solve it, I'll need to go deeper in the copying logic.

Update 3: There were more instances of /snapshot/ in the code, so I've replaced them with the new regex. Trying again to see if that fixes the issues. Unfortunately, the process is quite slow and painful. Added more logs for clarity and debugging.

Update 4: It worked!

Related https://github.com/prisma/ecosystem-tests/pull/3462

### CODE REVIEW

File: packages/engines/src/index.ts [hunk 1]
```diff
@@ -68,6 +69,18 @@ path.join(__dirname, '../query-engine-rhel-openssl-1.0.x')
...
+path.join(__dirname, '../schema-engine-darwin-arm64')
+path.join(__dirname, '../schema-engine-debian-openssl-1.0.x')
+path.join(__dirname, '../schema-engine-debian-openssl-1.1.x')
+path.join(__dirname, '../schema-engine-debian-openssl-3.0.x')
+path.join(__dirname, '../schema-engine-linux-static-x64')
+path.join(__dirname, '../schema-engine-linux-static-arm64')
+path.join(__dirname, '../schema-engine-rhel-openssl-1.0.x')
+path.join(__dirname, '../schema-engine-rhel-openssl-1.1.x')
+path.join(__dirname, '../schema-engine-rhel-openssl-3.0.x')
+
```
Comment (millsp) [inline packages/engines/src/index.ts:83]: I don't think this is needed in the end.

File: packages/internals/src/resolveBinary.ts [hunk 2]
```diff
@@ -23,10 +24,13 @@ async function getBinaryName(name: BinaryType): Promise<string> {
 
 export async function resolveBinary(name: BinaryType, proposedPath?: string): Promise<string> {
   // if file exists at proposedPath (and does not start with `/snapshot/` (= pkg), use that one
-  if (proposedPath && !proposedPath.startsWith('/snapshot/') && fs.existsSync(proposedPath)) {
+  if (proposedPath && !proposedPath.match(vercelPkgPathRegex) && fs.existsSync(proposedPath)) {
+    console.log(`resolveBinary: Using ${name} at proposedPath ${proposedPath}`)
```
Comment (millsp) [inline packages/internals/src/resolveBinary.ts:28]:
```suggestion
```

File: packages/internals/src/resolveBinary.ts [hunk 3]
```diff
@@ -23,10 +24,13 @@ async function getBinaryName(name: BinaryType): Promise<string> {
...
     return proposedPath
   }
 
+  console.log('resolveBinary: Proposed path not found, searching for', name)
+
```
Comment (millsp) [inline packages/internals/src/resolveBinary.ts:33]:
```suggestion
```

File: packages/internals/src/resolveBinary.ts [hunk 4]
```diff
@@ -38,24 +42,28 @@ export async function resolveBinary(name: BinaryType, proposedPath?: string): Pr
 
   const prismaPath = path.join(getEnginesPath(), binaryName)
   if (fs.existsSync(prismaPath)) {
+    console.log(`resolveBinary: 1. Found ${name} at ${prismaPath}`)
```
Comment (millsp) [inline packages/internals/src/resolveBinary.ts:45]:
```suggestion
```

File: packages/internals/src/resolveBinary.ts [hunk 5]
```diff
@@ -38,24 +42,28 @@ export async function resolveBinary(name: BinaryType, proposedPath?: string): Pr
...
     return maybeCopyToTmp(prismaPath)
   }
 
   // for pkg (related: https://github.com/vercel/pkg#snapshot-filesystem)
   const prismaPath2 = path.join(__dirname, '..', binaryName)
   if (fs.existsSync(prismaPath2)) {
+    console.log(`resolveBinary: 2. Found ${name} at ${prismaPath}`)
```
Comment (millsp) [inline packages/internals/src/resolveBinary.ts:52]:
```suggestion
```

File: packages/internals/src/resolveBinary.ts [hunk 6]
```diff
@@ -38,24 +42,28 @@ export async function resolveBinary(name: BinaryType, proposedPath?: string): Pr
...
     return maybeCopyToTmp(prismaPath2)
   }
 
   // TODO for ??
   const prismaPath3 = path.join(__dirname, '../..', binaryName)
   if (fs.existsSync(prismaPath3)) {
+    console.log(`resolveBinary: 3. Found ${name} at ${prismaPath}`)
```
Comment (millsp) [inline packages/internals/src/resolveBinary.ts:59]:
```suggestion
```

File: packages/internals/src/resolveBinary.ts [hunk 7]
```diff
@@ -38,24 +42,28 @@ export async function resolveBinary(name: BinaryType, proposedPath?: string): Pr
...
     return maybeCopyToTmp(prismaPath3)
   }
 
   // TODO for ?? / needed to come from @prisma/client/generator-build to @prisma/client/runtime
   const prismaPath4 = path.join(__dirname, '../runtime', binaryName)
   if (fs.existsSync(prismaPath4)) {
+    console.log(`resolveBinary: 4. Found ${name} at ${prismaPath}`)
```
Comment (millsp) [inline packages/internals/src/resolveBinary.ts:66]:
```suggestion
```

File: packages/internals/src/resolveBinary.ts [hunk 8]
```diff
@@ -79,7 +87,8 @@ export function safeResolveBinary(name: BinaryType, proposedPath?: string): TE.T
 export async function maybeCopyToTmp(file: string): Promise<string> {
   const dir = eval('__dirname')
 
-  if (dir.startsWith('/snapshot/')) {
+  if (dir.match(vercelPkgPathRegex)) {
+    console.log('maybeCopyToTpm: Vercel pkg detected')
```
Comment (millsp) [inline packages/internals/src/resolveBinary.ts:91]:
```suggestion
```

File: packages/engines/src/index.ts (hunk 1 above)
Comment (millsp) [inline packages/engines/src/index.ts:83]:
```suggestion
```

File: packages/engines/src/index.ts [hunk 9]
```diff
@@ -57,6 +57,7 @@ export { enginesVersion } from '@prisma/engines-version'
  * And needed for https://github.com/vercel/pkg#detecting-assets-in-source-code
  */
 
+path.join(__dirname, '../query-engine-windows.exe')
```
Comment (millsp) [inline packages/engines/src/index.ts:60]:
```suggestion
```

File: packages/engines/src/index.ts (hunk 1 above)
Comment (millsp) [reply packages/engines/src/index.ts]: Was there a reason these were not included though? The comment above says it's for `pkg`, so I found that counter-intuitive.

File: packages/fetch-engine/src/download.ts [hunk 10]
```diff
@@ -25,6 +25,9 @@ const exists = promisify(fs.exists)
 
 const channel = 'master'
 
+// matches `/snapshot/` or `C:\\snapshot\\` or `C:/snapshot/` for vercel's pkg apps
+export const vercelPkgPathRegex = /^((\w:[\\\/])|\/)snapshot[\/\\]/
```
Comment (millsp) [inline packages/fetch-engine/src/download.ts:29]: (Related [thoughts](https://prisma-company.slack.com/archives/C050YBB3BC4/p1688588505401419) about de-duplication)

**Review by Jolg42 — APPROVED** [review]

**Review by janpio — APPROVED** [review]

File: packages/engines/src/index.ts (hunk 1 above)
Comment (Jolg42) [reply packages/engines/src/index.ts]: Might not be needed any more indeed, some tech debt for another day.

**Pull request merged**